### DISCUSSION TITLE
How do I publish an npm package using tree-sitter that can be used in the browser?

### DISCUSSION THREAD

**Discussion by verhovsky** [tree-sitter/tree-sitter#1415] [Q&A]
I am writing a JS package that transpiles Bash commands into programs in other languages. My package is [one function](https://github.com/NickCarneiro/curlconverter/blob/547816a9f108e6953f1cc4f337a9fe837d802db9/util.js#L931) that takes a string of Bash source code and returns an object representing the arguments passed to the command that the Bash source code contains and then a series of functions that take that object and convert it into a string containing a program in a different programming language. It also [exports a Bash command](https://github.com/NickCarneiro/curlconverter/blob/547816a9f108e6953f1cc4f337a9fe837d802db9/package.json#L48) that runs that function on whatever a user pipes in from stdin. I want to use this package in the browser as well as from the command line (i.e. through Node).

I am having trouble understanding how to package this. I have two questions:

1) The README [says](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md#running-wasm-in-nodejs) 

> executing .wasm files in node.js is considerably slower than running node.js bindings

, but I am trying to create a package that can be used as both. Am I expected to make the people using my package in Node run the slower version or is there a way for users of my package that uses tree-sitter to conditionally tell it to use the Node.js bindings (`npm install tree-sitter`) instead of the WASM version (`npm install web-tree-sitter`)? Do I have to publish two packages `my-package` and `node-my-package` which only differ in which tree-sitter they use?

2) How do I include the WASM file in such a way that it can be found when running my package in the browser? I compiled `tree-sitter-bash` into a .wasm file myself and ship it in the package and unlike [the example in the README](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md#basic-usage), I can't just do

```
const Bash = await Parser.Language.load('./tree-sitter-bash.wasm');
parser.setLanguage(Bash);
```

because if I run that code with a different cwd (which happens when using the my package's exported command line tool) then the file isn't in the cwd. So as far as I can tell, you're supposed to do this:

```js
import path from 'path'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const bashGrammarFile = path.resolve(__dirname, 'tree-sitter-bash.wasm')

await Parser.init()
const Bash = await Parser.Language.load(bashGrammarFile)
const parser = new Parser()
parser.setLanguage(Bash)
```

but `fileURLToPath` doesn't exist in the browser...

**Comment by verhovsky** [comment]
After some hours of googling and reading, I think the answer to my first question is package.json's [`browser` field](https://github.com/defunctzombie/package-browser-field-spec). Basically I should have two separate files, `bashParser.js` and `bashParserBrowser.js` that import `tree-sitter` and `web-tree-sitter` respectively and initialize the parser object, and then I can tell bundlers like Webpack to replace `bashParser.js` with the browser version (when they're bundling my package for a browser) by putting this in my package.json:

```json
    "browser": {
        "./bashParser.js": "./bashParserBrowser.js"
    }
```

But I still don't know how I can pass the location of `tree-sitter-bash.wasm` to `await Parser.Language.load('tree-sitter-bash.wasm')` in `bashParserBrowser.js` when I don't know where it will be served from by people using my package. I came across this set of PRs/Issues/commits

https://github.com/tree-sitter/tree-sitter/pull/569

https://github.com/tree-sitter/tree-sitter/pull/1367

https://github.com/microsoft/vscode-anycode/commit/1a6b8818177635cb911e8f085ccf03219992a320

where people seem to be dealing with a similar question but being able to pass a `locateFile` function doesn't help me because I still don't understand what to put in that function.

**Reply by ahelwer** [reply]
idk whether this is useful but I've seen some projects that assume the wasm file will be in the root directory of wherever the tree-sitter grammar's NPM package is downloaded. That's where I put it in my grammar that I publish.

**Reply by verhovsky** [reply]
My package uses tree-sitter for Bash parsing. I don't want people using my package to know or care what it uses. Using tree-sitter requires serving two WASM files to the browser and now I have to explain to users what tree-sitter is and that they need to serve these files and why. It would be nice if they could just `npm install`, `import mypackage from 'mypackage'` and that's it. This is how all the other libraries I've used in the browser so far work, they don't make me learn about their internals and make me host files that they depend on.

What I ended up doing (as the first user of my own package), is [telling Webpack to copy the WASM files from `node_modules/`](https://github.com/verhovsky/curltorequests/blob/715ec4defd2ca661aa74fc9400e1fff9a183755b/webpack.config.js#L29-L32)

```js
  plugins: [
    new CopyPlugin({
      patterns: [
        'node_modules/web-tree-sitter/tree-sitter.wasm',
        'node_modules/<my package>/tree-sitter-bash.wasm',
      ]
    })
  ]
```

which feels like a hack. And again, I have to explain to anyone using my package that they need to do this.
